pkg/agent/util/net.go

```diff
@@ -106,47 +108,77 @@ func dialUnix(address string) (net.Conn, error) {
 	return net.Dial("unix", address)
 }
 
-// GetIPNetDeviceFromIP returns a local IP/mask and associated device from IP.
-func GetIPNetDeviceFromIP(localIP net.IP) (*net.IPNet, *net.Interface, error) {
+// GetIPNetDeviceFromIP returns local IPs/masks and associated device from IP.
+func GetIPNetDeviceFromIP(localIPs *ip.DualStackIPs) (v4IPNet *net.IPNet, v6IPNet *net.IPNet, iface *net.Interface, err error) {
 	linkList, err := net.Interfaces()
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
-	for _, link := range linkList {
-		addrList, err := link.Addrs()
+	// localIPs includes at most one IPv4 address and one IPv6 address. For each device in linkList, all its addresses
+	// are compared with IPs in localIPs. If found, the iface is set to the device and v4IPNet, v6IPNet are set to
+	// the matching addresses.
+	saveIface := func(current *net.Interface) error {
+		if iface != nil && iface.Index != current.Index {
+			return fmt.Errorf("IPs of localIPs should be on the same device")
```

Comment:
L148 will break the loop immediately once a link is found, how can this error happen?

Reply:
Removed `break` in the loop.